lib/router/src/cache.rs

```diff
@@ -1,426 +1,476 @@
-///
-/// Wraps a mutable reference to a `V`-typed value.
-///
-/// When the guard is dropped, the value's `last_access` time is updated with the provided
-/// time source.
-#[derive(Debug)]
-pub struct Access<'a, T: 'a, N: Now + 'a = ()> {
-    node: &'a mut Node<T>,
-    now: &'a N,
+#[derive(Debug, PartialEq)]
+pub struct CapacityExhausted {
+    pub(crate) capacity: usize,
 }
 
-/// A handle to a `Cache` that has capacity for at least one additional value.
-#[derive(Debug)]
-pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a, N: 'a> {
+/// A handle to a cache that has capacity for at least one additional value.
+pub struct Reserve<'a, K: Clone + Eq + Hash, V> {
+    expirations: &'a mut DelayQueue<K>,
+    expires: Duration,
     vals: &'a mut IndexMap<K, Node<V>>,
-    now: &'a N,
-}
-
-#[derive(Clone, Debug, PartialEq)]
-pub struct CapacityExhausted {
-    pub capacity: usize,
 }
 
 // ===== impl Cache =====
 
-impl<K: Hash + Eq, V> Cache<K, V, ()> {
-    pub fn new(capacity: usize, max_idle_age: Duration) -> Self {
-        Self {
+impl<K: Clone + Eq + Hash, V> Cache<K, V> {
+    pub fn new(capacity: usize, expires: Duration) -> (Lock<Cache<K, V>>, PurgeCache<K, V>) {
+        assert!(capacity != 0);
+        let cache = Self {
             capacity,
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
             vals: IndexMap::default(),
-            max_idle_age,
-            now: (),
-        }
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
+
+        (cache, bg_purge)
+    }
+
+    pub fn access(&mut self, key: &K) -> Option<Access<K, V>> {
+        let node = self.vals.get_mut(key)?;
+        Some(Access {
+            expires: self.expires,
+            expirations: &mut self.expirations,
+            node,
+        })
     }
-}
 
-impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
-    /// Accesses a route.
-    ///
-    /// A mutable reference to the route is wrapped in the returned `Access` to
-    /// ensure that the access-time is updated when the reference is released.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
-    where
-        Q: Hash + Equivalent<K>,
-    {
-        let v = self.vals.get_mut(key)?;
-        Some(v.access(&self.now))
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
 
-    /// Ensures that there is capacity to store an additional route.
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
+    pub fn reserve(&mut self) -> Async<Reserve<K, V>> {
         if self.vals.len() == self.capacity {
```

Comment:
Does this mean that we only try to evict items when the router is at capacity?

Reply:
Yes, but this only applies when trying to reserve a slot. The eager eviction comes from spawning the background task that will poll the DelayQueue without checking current length first. If we don’t spawn the background task, then the only way values are dropped is right here.